### PR TITLE
Support usernames recovery when claims are not unique

### PR DESCRIPTION
**Pull request by KD23243**
### Proposed changes in this pull request

- Implement a feature that allows users to recover multiple usernames associated with the same claims during the recovery process.
- Send multiple account recovery emails with all the usernames linked to the provided email address to help users identify their accounts.

### Follow up actions

- Update relevant documentation and user guides to explain how the new recovery process works.

### Related Git Issue
- https://github.com/wso2/product-is/issues/21073

### Related PRs
- https://github.com/wso2/carbon-identity-framework/pull/5946

#### General

- [x] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [x] **Is the PR labeled correctly?**

#### Functionality

- [x] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [x] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [x] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.

#### Tests

- [x] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [x] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [x] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [x] **Are all UI and API inputs run through forms or serializers?**
- [x] **Are all external inputs validated and sanitized appropriately?**
- [x] **Does all branching logic have a default case?**
- [x] **Does this solution handle outliers and edge cases gracefully?**
- [x] **Are all external communications secured and restricted to SSL?**

### CODE REVIEW

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java [hunk 1]
```diff
@@ -360,6 +358,83 @@ public String getUsernameByClaims(Map<String, String> claims, String tenantDomai
         }
     }
 
+    /**
+     * Get the user list for the given claims.
+     *
+     * @param claims       List of UserClaims
+     * @param tenantDomain Tenant domain
+     * @return Username (Return null if there are no users).
```
Comment (Malith-19) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java:366]: this returns user list right?

File: components/org.wso2.carbon.identity.recovery/src/test/java/org/wso2/carbon/identity/recovery/internal/service/impl/username/UsernameRecoveryManagerImplTest.java [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+    /**
+     * Mock the IdentityEventService.
+     */
+    private void mockIdentityEventService() {
+
+        mockedIdentityRecoveryServiceDataHolder.when(IdentityRecoveryServiceDataHolder::getInstance).thenReturn(
+                identityRecoveryServiceDataHolder);
+        when(identityRecoveryServiceDataHolder.getIdentityEventService()).thenReturn(identityEventService);
+    }
+}
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.recovery/src/test/java/org/wso2/carbon/identity/recovery/internal/service/impl/username/UsernameRecoveryManagerImplTest.java:413]: Add a new line here.

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/username/UsernameRecoveryManagerImpl.java [hunk 3]
```diff
@@ -187,18 +190,7 @@ public UsernameRecoverDTO notify(String recoveryCode, String channelId, String t
...
-            } catch (NumberFormatException e) {
-                if (log.isDebugEnabled()) {
-                    String message = String.format("Invalid boolean value : %s to enable legacyAPIs", properties
-                            .get(IdentityRecoveryConstants.USE_LEGACY_API_PROPERTY_KEY));
-                    log.debug(message);
-                }
-            }
-        }
-        return false;
+        return Boolean.parseBoolean(properties.get(IdentityRecoveryConstants.USE_LEGACY_API_PROPERTY_KEY));
```
[the remark this answers is not in the document]
Comment (ashensw) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/username/UsernameRecoveryManagerImpl.java:193]: Also better to keep the debug logs.

**Comment by KD23243** [comment]
The changed functionality has been tested with **WSO2IS-7.1.0-m2**. 

Both flows (with and without the configuration) are working as expected. For unique users, the functionality remains the same, and recovery emails are sent to their registered addresses. When the configuration is enabled, the system also sends multiple recovery notifications for non-unique users. Hence the behaviour matching the expected functionality.

**Review by ashensw — APPROVED** [review]

File: components/org.wso2.carbon.identity.recovery/src/test/java/org/wso2/carbon/identity/recovery/internal/service/impl/username/UsernameRecoveryManagerImplTest.java [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+                .thenReturn(new IdentityRecoveryClientException(null));
+        usernameRecoveryManager.notify(recoveryCode, "2", TENANT_DOMAIN, properties);
+    }
+
+    /**
+     * Test to get callback URL with exception.
+     *
+     * @throws IdentityRecoveryException if an error occurs during retrieval.
+     */
+    @Test(expectedExceptions = NullPointerException.class)
```
Comment (sadilchamishka) [inline components/org.wso2.carbon.identity.recovery/src/test/java/org/wso2/carbon/identity/recovery/internal/service/impl/username/UsernameRecoveryManagerImplTest.java:331]: From where the NPE is thrown. Ideally we should avoid NPE. The test should cover the failure.

Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.recovery/src/test/java/org/wso2/carbon/identity/recovery/internal/service/impl/username/UsernameRecoveryManagerImplTest.java:331]: Check other test cases also written for NPE

File: components/org.wso2.carbon.identity.recovery/src/test/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManagerTest.java [hunk 5]
```diff
@@ -157,6 +175,667 @@ public void testRetrieveUserRecoveryInformation() throws Exception {
...
     }
 
+    /**
+     * Tests that a NullPointerException is thrown during user recovery when a UserStoreException
+     * occurs.
+     *
+     * @throws Exception if mocking or method invocation fails.
+     */
+    @Test
+    public void testThrowNullPointerForUserRecovery() throws Exception {
```
Comment (PasinduYeshan) [inline components/org.wso2.carbon.identity.recovery/src/test/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManagerTest.java:185]: Why are we throwing NPE from UserStoreException. NPEs typically indicate programming errors, Ideally, method should catch the UserStoreException and throw a more specific checked exception or throw the error it self.

File: components/org.wso2.carbon.identity.recovery/src/test/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManagerTest.java [hunk 6]
```diff
@@ -157,6 +175,667 @@ public void testRetrieveUserRecoveryInformation() throws Exception {
...
+        userAccountRecoveryManager.retrieveUserRecoveryInformation(userClaims, StringUtils.EMPTY,
+                RecoveryScenarios.USERNAME_RECOVERY, properties);
+    }
+
+    /**
+     * Tests testGetDomainNames through retrieveUserRecoveryInformation
+     *
+     * @throws Exception if there is an issue with mocking or method invocation.
+     */
+    @Test(expectedExceptions = NullPointerException.class)
```
Comment (PasinduYeshan) [inline components/org.wso2.carbon.identity.recovery/src/test/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManagerTest.java:773]: Is NPE intentional?

Comment (PasinduYeshan) [reply components/org.wso2.carbon.identity.recovery/src/test/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManagerTest.java:773]: https://github.com/wso2-extensions/identity-governance/pull/859/files#r2041072832

**Pull request merged**